### PR TITLE
Validate df schema for graph.construct

### PR DESCRIPTION
**Pull request by FlorentinD**

### CODE REVIEW

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        return license == "Licensed"
+
+    def _node_query(self, node_df: DataFrame) -> str:
+        node_id_index = node_df.columns.get_loc("nodeId")
+
+        label_query = ""
+        if "labels" in node_df.keys():
+            label_index = node_df.columns.get_loc("labels")
+            label_query = f", node[{label_index}] as labels"
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:68]:
```suggestion
            label_query = f", node[{label_index}] AS labels"
```

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+        property_query = ""
+        property_columns: Set[str] = set(node_df.keys()) - {"nodeId", "labels"}
+        if len(property_columns) > 0:
+            property_queries = (f", node[{node_df.columns.get_loc(col)}] as {col}" for col in property_columns)
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:73]:
```suggestion
            property_queries = (f", node[{node_df.columns.get_loc(col)}] AS {col}" for col in property_columns)
```

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            property_query = "".join(property_queries)
+
+        return f"UNWIND $nodes as node RETURN node[{node_id_index}] as id{label_query}{property_query}"
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:76]:
```suggestion
        return f"UNWIND $nodes AS node RETURN node[{node_id_index}] AS id{label_query}{property_query}"
```

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    def _relationship_query(self, rel_df: DataFrame) -> str:
+        source_id_index = rel_df.columns.get_loc("sourceNodeId")
+        target_id_index = rel_df.columns.get_loc("targetNodeId")
+
+        type_query = ""
+        if "relationshipType" in rel_df.keys():
+            type_index = rel_df.columns.get_loc("relationshipType")
+            type_query = f", relationship[{type_index}] as type"
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:85]:
```suggestion
            type_query = f", relationship[{type_index}] AS type"
```

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+        property_query = ""
+        property_columns: Set[str] = set(rel_df.keys()) - {"sourceNodeId", "targetNodeId", "relationshipType"}
+        if len(property_columns) > 0:
+            property_queries = (f", relationship[{rel_df.columns.get_loc(col)}] as {col}" for col in property_columns)
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:90]:
```suggestion
            property_queries = (f", relationship[{rel_df.columns.get_loc(col)}] AS {col}" for col in property_columns)
```

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            property_query = "".join(property_queries)
+
+        return (
+            "UNWIND $relationships as relationship "
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:94]:
```suggestion
            "UNWIND $relationships AS relationship "
```

File: graphdatascience/query_runner/cypher_graph_constructor.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            f"RETURN relationship[{source_id_index}] as source, relationship[{target_id_index}] as target"
```
Comment (knutwalker) [inline graphdatascience/query_runner/cypher_graph_constructor.py:95]:
```suggestion
            f"RETURN relationship[{source_id_index}] AS source, relationship[{target_id_index}] AS target"
```

File: graphdatascience/tests/unit/test_graph_ops.py [hunk 8]
```diff
@@ -324,3 +324,57 @@ def test_graph_generate(runner: CollectingQueryRunner, gds: GraphDataScience) ->
...
+        }
+    )
+
+    gds.alpha.graph.construct("hello", nodes, relationships, concurrency=2)
+
+    expected_node_query = "UNWIND $nodes as node RETURN node[0] as id, node[1] as labels, node[2] as propA"
+    expected_relationship_query = (
+        "UNWIND $relationships as relationship RETURN "
+        "relationship[0] as source, relationship[1] as target, "
+        "relationship[2] as type, relationship[3] as relPropA"
```
Comment (knutwalker) [inline graphdatascience/tests/unit/test_graph_ops.py:353]:
```suggestion
    expected_node_query = "UNWIND $nodes AS node RETURN node[0] AS id, node[1] AS labels, node[2] AS propA"
    expected_relationship_query = (
        "UNWIND $relationships AS relationship RETURN "
        "relationship[0] AS source, relationship[1] AS target, "
        "relationship[2] AS type, relationship[3] AS relPropA"
```

**Review by knutwalker — APPROVED** [review]
looks good, just some suggestions to make @jjaderberg happy :)
I think I also reviewed the base PR, maybe because this PR wasn't rebased after the base PR got merged?

File: graphdatascience/graph/graph_proc_runner.py [hunk 9]
```diff
@@ -267,13 +267,25 @@ def construct(
...
+            if "nodeId" not in node_df.columns.values:
+                errors.append(f"Node dataframe at index {idx} needs to contain a 'nodeId' column.")
+
+        for idx, rel_df in enumerate(relationships):
+            for expected_col in ["sourceNodeId", "targetNodeId"]:
+                if expected_col not in rel_df.columns.values:
+                    errors.append(f"Relationship dataframe at index {idx} needs to contain a '{expected_col}' column.")
+
+        if len(errors) > 0:
+            raise ValueError(f"{os.linesep}".join(errors))
```
[the remark this answers is not in the document]
Comment (adamnsch) [reply graphdatascience/graph/graph_proc_runner.py]: Done

**Pull request merged**